crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1563,8 +1785,153 @@ impl<'db> Type<'db> {
         arguments: &CallArguments<'_, 'db>,
     ) -> Result<CallOutcome<'db>, CallError<'db>> {
         match self {
+            Type::Callable(CallableType::BoundMethod(bound_method)) => {
+                let instance = bound_method.self_instance(db);
+                let arguments = arguments.with_self(instance);
+
+                let binding = bind_call(
+                    db,
+                    &arguments,
+                    bound_method.function(db).signature(db),
+                    self,
+                );
+
+                if binding.has_binding_errors() {
+                    Err(CallError::BindingError { binding })
+                } else {
+                    Ok(CallOutcome::Single(binding))
+                }
+            }
+            Type::Callable(CallableType::MethodWrapperDunderGet(function)) => {
+                // Here, we dynamically model the overloaded function signature of `types.FunctionType.__get__`.
+                // This is required because we need to return more precise types than what the signature in
+                // typeshed provides:
+                //
+                // ```py
+                // class FunctionType:
+                //     # ...
+                //     @overload
+                //     def __get__(self, instance: None, owner: type, /) -> FunctionType: ...
+                //     @overload
+                //     def __get__(self, instance: object, owner: type | None = None, /) -> MethodType: ...
+                // ```
+
+                let first_argument_is_none =
+                    arguments.first_argument().is_some_and(|ty| ty.is_none(db));
+
+                let signature = Signature::new(
+                    Parameters::new([
+                        Parameter::new(
+                            Some("instance".into()),
+                            Some(Type::object(db)),
+                            ParameterKind::PositionalOnly { default_ty: None },
+                        ),
+                        if first_argument_is_none {
+                            Parameter::new(
+                                Some("owner".into()),
+                                Some(KnownClass::Type.to_instance(db)),
+                                ParameterKind::PositionalOnly { default_ty: None },
+                            )
+                        } else {
+                            Parameter::new(
+                                Some("owner".into()),
+                                Some(UnionType::from_elements(
+                                    db,
+                                    [KnownClass::Type.to_instance(db), Type::none(db)],
+                                )),
+                                ParameterKind::PositionalOnly {
+                                    default_ty: Some(Type::none(db)),
+                                },
+                            )
+                        },
+                    ]),
```

Comment:
the construction of these parameter types looks like it could possibly be cached as a Salsa query? That _might_ improve the perf here a bit?

Simply cloning a short array of `Type`s might be more performant than constructing a new one from scratch each time, given that constructing a new one from scratch requires many Salsa db lookups, but cloning a `Type` should be cheap since most variants are just a u32